src/FileHistory.cc

```diff
@@ -256,6 +259,9 @@ QVariant FileHistory::data(const QModelIndex& index, int role) const {
   if (col == QGit::LOG_COL)
     return r->shortLog();
 
+  if (col == QGit::HASH_COL)
+    return r->shortHash(8);
```

Comment:
Please see https://stackoverflow.com/questions/18134627/how-much-of-a-git-sha-is-generally-considered-necessary-to-uniquely-identify-a. A short SHA of 8 will be problematic when loading the kernel git clone or the git git clone in qgit.  Please either increase the magic constant to 12, or add a brief algorithm that selects the proper length from the number of objects in the current project.

Reply:
The only accurate algorithm would be to compare the hashes of all commits, no?  Any quick rule of thumb method would fail eventually.  Some quick to implement alternatives are to make this a preference or just always show all the digits.  Personally, I could live with hardcoding 12.

I see a preference was added to show the short version in the upper-right line edit and that currently shows seven digits.

Reply:
Hardcoding to 12 is OK with me too.

Just for the sake of the interesting issue: the goal is _not_ to have a precise algorithm (which would indeed be O(n) on the number of commits, thus, as you hint, inconvenient for UI speed. The goal is to make collisions very improbable (as explained in the post I linked, the "birthday problem" mention towards the end). I.e., select a length of short SHA, with the small probability formulae given in that post, that would make it very _improbable_ that there be a collision.

Note that even the _full_ (40 chars) SHA does _not_ guarantee total absence of collisions. Its length just makes them extremely improbable. That's the beauty of hashes: they are fast to calculate, thus avoiding otherwise necessary full enumerations. The tradeoff is a very small probability of collision between keys.  

Again, thank you very much.

Reply:
You are right when you imply that the already existing "short hash" will behave now in an  anachronic manner. I had forgotten about that. Checking the code, that uses `git rev-parse --short`, which is a bit contrived, but it _will_ give the appropriate SHA length depending on the repository side. It's just so happens that for some majority of repos, 7-8 is enough.

Reply:
That git command is run each time the QLineEdit changes, which is not very efficient.  If the command was run once on each Refresh View and the length stored, that value could be used for both the Short Hash column and the QLineEdit.

Reply:
This would be one solution.

A preferable one would be to calculate the appropriate size of the short SHA once at the load of the repository, store it in a globally accessible variable, and reuse it both for your new SHA column feature and for the previous "short sha in toolbar" feature.
This constant can even be obtained by calling `git rev-parse --short` only once, on any of the repo's commit SHAs, if we don't want to implement small the collision probability calculation algorithm.

Reply:
Hmm... using that "Display short commit hash" option, I just got the following error dialog when I applied a log search filter.  Perhaps the sha string is empty for some reason.  The change suggested above may fix this bug.

> An error occurred while executing command:
> git rev-parse --short
> fatal: Needed a single revision

Reply:
You've just repeated the proposal.  Where is the best place to hook into the repository refresh and where's the best place to store global repository statistics?

Reply:
At first blush, I'd add the initial calculation in MainImpl::setRepository() and I'd store the result in a MainImpl attribute, exposed through a new public method.